task3/task3_solution.py

```diff
@@ -0,0 +1,150 @@
+        return res.strip()
+
+    # calc overall average grade
+    def avg_grade(self, grades):
+        res = sum(grades)/len(grades)
+        return res
+
+    def inDict(self, name):
+        found = False
+        for dict in self.student_data:
+            if name in dict.values():
+                found = True
+                return found
+                break
+        return found
+
+    def get_letter_grades(self, *students, throw_exception=False):
+        for name in students:
+            # if name is present in dictionaries, do stuff
+            if self.inDict(name):
+                for dict in self.student_data:
+                    if dict.get("name") == name:
+                        print("Grades of " + name + ": ")
+                        temp = dict.copy()
+                        temp.pop("name")
+                        for work, grades in temp.items():
+                            print(work, self.to_letters(grades))
+
+                        # obsolete code, for testing purposes
+                        #
+                        # print("Homework:", self.to_letters(dict["homework"]))
+                        # print("Quizzes:", self.to_letters(dict["quizzes"]))
+                        # print("Tests:", self.to_letters(dict["tests"]))
+            else:
+                # if keyword specified, show error instead of skipping
+                if throw_exception:
+                    print("No entry", name, "in table")
+
+    def get_student_average(self, *students, throw_exception=False):
+        for name in students:
+            # if name is present in dictionaries, do stuff
+            if self.inDict(name):
+                for dict in self.student_data:
+                    if dict.get("name") == name:
+                        temp = dict.copy()
+                        temp.pop("name")
+                        # there might be an overkill here or an easier way
+                        # to pass list of values to avg_grade
+                        list = []
+                        for values in temp.values():
+                            list.extend(values)
+                        print("Average grade of", name, ":",
+                              "{0:5.2f}".format(self.avg_grade(list)))
+
+            else:
+                # if keyword specified, show error instead of skipping
+                if throw_exception:
+                    print("No entry", name, "in table")
+
+    # to see the state of data
+    def test(self):
+        for data in self.student_data:
+            print([value for key, value in data.items()])
+
+    def update_grade(self, student_name, throw_exception=False, **kwargs):
+        if self.inDict(student_name):
+            for dict in self.student_data:
+                # seems like im breaking DRY next line
+                if dict.get("name") == student_name:
+                    for new_work, new_grades in kwargs.items():
+                        for work, grades in dict.items():

```

Comment:
You can just do `dict[new_work].extend(new_grades)`

Reply:
And overall main reason to use dicts is that you can just grab stuff using key with O(1), looping over them usually can be avoided with `.get()` or some other built in method

Reply:
yes, good point. I've read about dict features and learnt that you don't always need to iterate, but I need to get a hand in using them. is it a good practice just to use as many expressions with `.get()` as you need? 
also I don't realize why `.get()` can't accept a variable (e.g. string or another hashable) as an argument, but rather takes a string directly - `.get("homework")` works fine, `.get(new_work)` doesn't. faulty code or a misconception?